packages/sources/galaxy/src/adapter.ts

```diff
@@ -0,0 +1,109 @@
+  Builder.selectEndpoint<Config, endpoints.TInputParameters>(request, makeConfig(), endpoints)
+
+export const makeExecute: ExecuteFactory<Config, endpoints.TInputParameters> = (config) => {
+  return async (request, context) => execute(request, context, config || makeConfig())
+}
+
+export const getAccessToken = async (config: Config): Promise<string> => {
+  try {
+    const tokenResponse = await Requester.request<AccessTokenResponse>({
+      url: config.api?.baseURL,
+      method: 'get' as Method,
+      headers: {
+        'X-GALAXY-APIKEY': (config.adapterSpecificParams?.apiKey || '') as string,
+        'X-GALAXY-PASSWORD': (config.adapterSpecificParams?.apiPassword || '') as string,
+      },
+    })
+    if (!tokenResponse.data.token)
+      throw new AdapterDataProviderError({ message: tokenResponse.data.message || 'Login failed' })
+    return tokenResponse.data.token
+  } catch (e: any) {
+    const err = e as any
+    const message = `Login failed ${err.message ? `with message '${err.message}'` : ''}`
+    const error = { ...err, message }
+    Logger.debug(message)
+    throw error.response
+      ? new AdapterDataProviderError(error)
+      : error.request
+      ? new AdapterConnectionError(error)
+      : new AdapterError(error)
+  }
+}
+
+export const makeWSHandler = (defaultConfig?: Config): MakeWSHandler<any> => {
+  let token: string | undefined
+  const getPair = (input: AdapterRequest) => {
+    const validator = new Validator(
+      input,
+      endpoints.price.inputParameters,
+      {},
+      { shouldThrowError: false },
+    )
+    const { base = '', quote = '' } = validator.validated.data
+    return base && quote && `markPrice_${base.toUpperCase()}/${quote.toUpperCase()}`
+  }
+
+  return async () => {
+    const config = defaultConfig || makeConfig()
+    if (!token) token = await getAccessToken(config)
+
+    return {
+      connection: {
+        url: config.ws?.baseWsURL,
+        protocol: { headers: { ...config.api?.headers, token } },
+      },
+      noHttp: true,
+      subscribe: (input: AdapterRequest) => ({ type: 'subscribe', signals: [getPair(input)] }),
+      unsubscribe: (input: AdapterRequest) => ({ type: 'unsubscribe', signals: [getPair(input)] }),
+      subsFromMessage: (message: TickerMessage, subscriptionMessage: Pair) => {
+        if (message.type !== 'signal_update') return
+        if (!subscriptionMessage.signals.includes(message.signal)) return
+        return { type: 'subscribe', signals: subscriptionMessage.signals }
+      },
+      isError: (message: TickerMessage) => message.type === 'error',
+      filter: (message: TickerMessage) => message.type === 'signal_update',
+      toResponse: (message: TickerMessage, input: AdapterRequest) => {
+        const result = Requester.validateResultNumber(message, ['value'])
+        return Requester.success(input.id, { data: { result, ts: message.ts } }, config.verbose)
+      },
+    } as any // TODO: connection type mismatch
```

Comment:
Connection type mismatch?

Reply:
Oh I see, it's the same issue as with GSR. I think these can be handled as a unit in a separate PR update.